content/tutorials/build-apps/build-a-desktop-wallet-in-javascript.md

```diff
@@ -0,0 +1,1441 @@
+[`view/4_preload.js`]({{target.github_forkurl}}/tree/{{target.github_branch}}/content/_code-samples/build-a-wallet/desktop-js/view/4_tx-preload.js).
+[`view/4_tx-history.html`]({{target.github_forkurl}}/tree/{{target.github_branch}}/content/_code-samples/build-a-wallet/desktop-js/view/4_tx-history.html).
+[`view/4_renderer.js`]({{target.github_forkurl}}/tree/{{target.github_branch}}/content/_code-samples/build-a-wallet/desktop-js/view/4_tx-renderer.js).
+
+At this point, our wallet shows the account's balance getting updated, but doesn't give us any clue about how this state came about, namely the actual transactions that caused the updates. So, our next step is to display the account's up to date transaction history using subscriptions once again:
+
+![Screenshot: Step 4, show transaction history](img/javascript-wallet-4.png)
+
+1. In the `library` folder, add a new file `4_helpers.js`. Then add the following helper function to that file:
+
+{{ include_code("_code-samples/build-a-wallet/desktop-js/library/4_helpers.js", language="js") }}
+
+2. Now, in `index.html`, require the new helper function at the bottom of the import section like so:
+3. 
+```javascript
+const { prepareReserve, prepareAccountData, prepareLedgerData} = require('./library/3_helpers')
+const { prepareTxData } = require('./library/4_helpers')
+```
+
+3. In the main file, update the listener function subscribed to the `transaction` event by adding the following snippet:
+
+```javascript
+// Wait for transaction on subscribed account and re-request account data
+client.on("transaction", async (transaction) => {
+  // Reference: https://xrpl.org/account_info.html
+  const accountInfoRequest = {
+    "command": "account_info",
+    "account": address,
+    "ledger_index": transaction.ledger_index
+  }
+
+  const accountInfoResponse = await client.request(accountInfoRequest)
+  const accountData = prepareAccountData(accountInfoResponse.result.account_data, reserve)
+  appWindow.webContents.send('update-account-data', accountData)
+
+  // Step 3 code additions - start
+  const transactions = prepareTxData([{tx: transaction.transaction}])
+  appWindow.webContents.send('update-transaction-data', transactions)
+  // Step 4 code additions - end
+  
+})
+```
+
+4. In `view/preload.js`, add the following code at the bottom of `exposeInMainWorld()`:
+
+```javascript
+onEnterAccountAddress: (address) => {
+  ipcRenderer.send('address-entered', address)
+},
+onUpdateAccountData: (callback) => {
+  ipcRenderer.on('update-account-data', callback)
+},
+
+// Step 4 code additions - start
+onUpdateTransactionData: (callback) => {
+  ipcRenderer.on('update-transaction-data', callback)
+}
+// Step 4 code additions - end
+```
+
+5. Modify `view/template.html` by adding a new fieldset below the ones that are already there:
+
+```html
+        ...
+        Close Time: <strong id="ledger-close-time"></strong><br/>
+    </fieldset>
+
+    <!-- Step 4 code additions - start -->
```

Comment:
```suggestion
    <!-- Step 5 code additions - start -->
```

Reply:
Mh, on this one the end was wrong, I have fixed it.